common/compositor/gl/glprogram.cpp

```diff
@@ -126,8 +126,93 @@ static std::string GenerateFragmentShader(int layer_count) {
   return fragment_shader_stream.str();
 }
 
+static GLint LoadPreBuiltBinary(GLint gl_program,
+                                void *binary, long size) {
+  GLint status;
+
+  std::string extensions((const char *)glGetString(GL_EXTENSIONS));
+  std::string mesa_prog_bin_fmts("GL_MESA_program_binary_formats");
+  std::string::size_type pos;
+
+  /* check if glProgramBinaryOES exists */
+  if (!glProgramBinaryOES)
+    return 0;
+
+  pos = extensions.find(mesa_prog_bin_fmts);
+
+  /* check GL_MESA_program_binary_formats exists */
+  if (pos == std::string::npos)
+    return 0;
+
+  /* TODO: we should check whether GL_MESA_program_binary_formats are
+   * supported here: however, this extension is not currently exposed now
+   * in MESA tip. Once it's fixed we will add a check routine here
+   */
+  glProgramBinaryOES(gl_program, GL_PROGRAM_BINARY_FORMAT_MESA, binary,
+                     size);
+
+  glGetProgramiv(gl_program, GL_LINK_STATUS, &status);
+  if (status)
+      return gl_program;
+
+  return 0;
+}
+
 static GLint GenerateProgram(unsigned num_textures,
                              std::ostringstream *shader_log) {
+  GLint status;
+  GLint program = glCreateProgram();
+  void *binary_prog;
+  long binary_sz;
+
+  if (!program) {
+    if (shader_log)
+      *shader_log << "Failed to create program."
+                  << "\n";
+    return 0;
+  }
+
+  /* try to load prebuilt shader program from files */
+  std::ostringstream shader_program_fname;
+  shader_program_fname << PREBUILT_SHADER_FILE_PATH"hwc_shader_prog_"
+                       << num_textures
+                       <<".shader_test.bin";
+
+  FILE *shader_prog_fp;
+
+  shader_prog_fp = fopen(shader_program_fname.str().c_str(), "rb");
+
+  if (shader_prog_fp) {
+    /* check the size of file */
+    fseek(shader_prog_fp, 0, SEEK_END);
+    binary_sz = ftell(shader_prog_fp);
+    rewind(shader_prog_fp);
+
+    char *binary_prog = (char *)malloc(binary_sz);
```

Comment:
I'd still like to see some bounds checking first. Come up with some maximum acceptable binary size and don't accept files that are bigger.

Reply:
We could do that. Is it ok to set it to 10MB for now?

Reply:
That sounds reasonable to me